File: crypto/tls/src/main/java/org/openecard/crypto/tls/SocketWrapper.java
/****************************************************************************
 * Copyright (C) 2013 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 ***************************************************************************/

package org.openecard.crypto.tls;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.Socket;
import java.net.SocketAddress;
import java.net.SocketException;
import java.nio.channels.SocketChannel;
import javax.annotation.Nonnull;


/**
 * Wrapper class to simulate socket for TLS in- and output streams.
 * BouncyCastle is only able to emit InputStream and OutputStream classes which represent the tunneled channel. If it is
 * desirable to operate on a socket, say because some API needs a socket, then this class can be used to bring the
 * tunneled streams and the originating socket together.
 *
 * @author Tobias Wich
 */
public class SocketWrapper extends Socket {

    private final Socket parent;
    private final InputStream in;
    private final OutputStream out;

    /**
     * Creates an instance of a SocketWrapper binding the given streams to the socket.
     * The socket must be opened and the streams must belong to the socket. The latter requirement is not checked.
     *
     * @param parent Connected socket which should be wrapped.
     * @param in Input stream belonging to the socket.
     * @param out Output stream belonging to the socket.
     * @throws IOException Thrown in case the socket is not connected.
     */
    public SocketWrapper(@Nonnull Socket parent, @Nonnull InputStream in, @Nonnull OutputStream out)
	    throws IOException {
	this.parent = parent;
	this.in = in;
	this.out = out;
	// assert that the socket is really open. An unconnected socket can not have open streams
	if (! parent.isConnected()) {
	    throw new IOException("Socket is not connected.");
	}
    }


    @Override
    public InputStream getInputStream() throws IOException {
	// call original to check if the socket is closed already
	parent.getInputStream();
	return in;
    }

    @Override
    public OutputStream getOutputStream() throws IOException {
	// call original to check if the socket is closed already
	parent.getInputStream();
	return out;
    }


    @Override
    public void bind(SocketAddress bindpoint) throws IOException {
	parent.bind(bindpoint);
    }

    @Override
    public synchronized void close() throws IOException {
	parent.close();
    }

    @Override
    public void connect(SocketAddress endpoint) throws IOException {
	parent.connect(endpoint);
    }

    @Override
    public void connect(SocketAddress endpoint, int timeout) throws IOException {
	parent.connect(endpoint, timeout);
    }

    @Override
    public SocketChannel getChannel() {
	return parent.getChannel();
    }

    @Override
    public InetAddress getInetAddress() {
	return parent.getInetAddress();
    }

    @Override
    public boolean getKeepAlive() throws SocketException {
	return parent.getKeepAlive(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public InetAddress getLocalAddress() {
	return parent.getLocalAddress(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public int getLocalPort() {
	return parent.getLocalPort(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public SocketAddress getLocalSocketAddress() {
	return parent.getLocalSocketAddress(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public boolean getOOBInline() throws SocketException {
	return parent.getOOBInline(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public int getPort() {
	return parent.getPort(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public synchronized int getReceiveBufferSize() throws SocketException {
	return parent.getReceiveBufferSize(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public SocketAddress getRemoteSocketAddress() {
	return parent.getRemoteSocketAddress(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public boolean getReuseAddress() throws SocketException {
	return parent.getReuseAddress(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public synchronized int getSendBufferSize() throws SocketException {
	return parent.getSendBufferSize(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public int getSoLinger() throws SocketException {
	return parent.getSoLinger(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public synchronized int getSoTimeout() throws SocketException {
	return parent.getSoTimeout(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public boolean getTcpNoDelay() throws SocketException {
	return parent.getTcpNoDelay(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public int getTrafficClass() throws SocketException {
	return parent.getTrafficClass(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public boolean isBound() {
	return parent.isBound(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public boolean isClosed() {
	return parent.isClosed(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public boolean isConnected() {
	return parent.isConnected(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public boolean isInputShutdown() {
	return parent.isInputShutdown(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public boolean isOutputShutdown() {
	return parent.isOutputShutdown(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void sendUrgentData(int data) throws IOException {
	parent.sendUrgentData(data); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void setKeepAlive(boolean on) throws SocketException {
	parent.setKeepAlive(on); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void setOOBInline(boolean on) throws SocketException {
	parent.setOOBInline(on); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void setPerformancePreferences(int connectionTime, int latency, int bandwidth) {
	parent.setPerformancePreferences(connectionTime, latency, bandwidth); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public synchronized void setReceiveBufferSize(int size) throws SocketException {
	parent.setReceiveBufferSize(size); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void setReuseAddress(boolean on) throws SocketException {
	parent.setReuseAddress(on); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public synchronized void setSendBufferSize(int size) throws SocketException {
	parent.setSendBufferSize(size); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void setSoLinger(boolean on, int linger) throws SocketException {
	parent.setSoLinger(on, linger); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public synchronized void setSoTimeout(int timeout) throws SocketException {
	parent.setSoTimeout(timeout); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void setTcpNoDelay(boolean on) throws SocketException {
	parent.setTcpNoDelay(on); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void setTrafficClass(int tc) throws SocketException {
	parent.setTrafficClass(tc); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void shutdownInput() throws IOException {
	parent.shutdownInput(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public void shutdownOutput() throws IOException {
	parent.shutdownOutput(); //To change body of generated methods, choose Tools | Templates.
    }

    @Override
    public String toString() {
	return parent.toString(); //To change body of generated methods, choose Tools | Templates.
    }

}

File: crypto/tls/src/jvmMain/kotlin/org/openecard/crypto/tls/SocketWrapper.kt
/****************************************************************************
 * Copyright (C) 2013 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 ***************************************************************************/
package org.openecard.crypto.tls

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.InetAddress
import java.net.Socket
import java.net.SocketAddress
import java.net.SocketException
import java.nio.channels.SocketChannel

/**
 * Wrapper class to simulate socket for TLS in- and output streams.
 * BouncyCastle is only able to emit InputStream and OutputStream classes which represent the tunneled channel. If it is
 * desirable to operate on a socket, say because some API needs a socket, then this class can be used to bring the
 * tunneled streams and the originating socket together.
 *
 * @author Tobias Wich
 */
class SocketWrapper(
	private val parent: Socket,
	private val `in`: InputStream,
	private val out: OutputStream,
) : Socket() {
	/**
	 * Creates an instance of a SocketWrapper binding the given streams to the socket.
	 * The socket must be opened and the streams must belong to the socket. The latter requirement is not checked.
	 *
	 * @param parent Connected socket which should be wrapped.
	 * @param `in` Input stream belonging to the socket.
	 * @param out Output stream belonging to the socket.
	 * @throws IOException Thrown in case the socket is not connected.
	 */
	init {
		// assert that the socket is really open. An unconnected socket can not have open streams
		if (!parent.isConnected) {
			throw IOException("Socket is not connected.")
		}
	}

	@Throws(IOException::class)
	override fun getInputStream(): InputStream {
		// call original to check if the socket is closed already
		parent.getInputStream()
		return `in`
	}

	@Throws(IOException::class)
	override fun getOutputStream(): OutputStream {
		// call original to check if the socket is closed already
		parent.getInputStream()
		return out
	}

	@Throws(IOException::class)
	override fun bind(bindpoint: SocketAddress?) {
		parent.bind(bindpoint)
	}

	@Synchronized
	@Throws(IOException::class)
	override fun close() {
		parent.close()
	}

	@Throws(IOException::class)
	override fun connect(endpoint: SocketAddress) {
		parent.connect(endpoint)
	}

	@Throws(IOException::class)
	override fun connect(
		endpoint: SocketAddress,
		timeout: Int,
	) {
		parent.connect(endpoint, timeout)
	}

	override fun getChannel(): SocketChannel? = parent.channel

	override fun getInetAddress(): InetAddress? = parent.getInetAddress()

	@Throws(SocketException::class)
	override fun getKeepAlive(): Boolean = parent.getKeepAlive()

	override fun getLocalAddress(): InetAddress = parent.getLocalAddress()

	override fun getLocalPort(): Int = parent.getLocalPort()

	override fun getLocalSocketAddress(): SocketAddress? = parent.getLocalSocketAddress()

	@Throws(SocketException::class)
	override fun getOOBInline(): Boolean = parent.getOOBInline()

	override fun getPort(): Int = parent.getPort()

	@Synchronized
	@Throws(SocketException::class)
	override fun getReceiveBufferSize(): Int = parent.getReceiveBufferSize()

	override fun getRemoteSocketAddress(): SocketAddress? = parent.getRemoteSocketAddress()

	@Throws(SocketException::class)
	override fun getReuseAddress(): Boolean = parent.getReuseAddress()

	@Synchronized
	@Throws(SocketException::class)
	override fun getSendBufferSize(): Int = parent.getSendBufferSize()

	@Throws(SocketException::class)
	override fun getSoLinger(): Int = parent.getSoLinger()

	@Synchronized
	@Throws(SocketException::class)
	override fun getSoTimeout(): Int = parent.getSoTimeout()

	@Throws(SocketException::class)
	override fun getTcpNoDelay(): Boolean = parent.getTcpNoDelay()

	@Throws(SocketException::class)
	override fun getTrafficClass(): Int = parent.trafficClass

	override fun isBound(): Boolean = parent.isBound

	override fun isClosed(): Boolean = parent.isClosed

	override fun isConnected(): Boolean = parent.isConnected

	override fun isInputShutdown(): Boolean = parent.isInputShutdown

	override fun isOutputShutdown(): Boolean = parent.isOutputShutdown

	@Throws(IOException::class)
	override fun sendUrgentData(data: Int) {
		parent.sendUrgentData(data)
	}

	@Throws(SocketException::class)
	override fun setKeepAlive(on: Boolean) {
		parent.setKeepAlive(on)
	}

	@Throws(SocketException::class)
	override fun setOOBInline(on: Boolean) {
		parent.setOOBInline(on)
	}

	override fun setPerformancePreferences(
		connectionTime: Int,
		latency: Int,
		bandwidth: Int,
	) {
		parent.setPerformancePreferences(connectionTime, latency, bandwidth)
	}

	@Synchronized
	@Throws(SocketException::class)
	override fun setReceiveBufferSize(size: Int) {
		parent.setReceiveBufferSize(size)
	}

	@Throws(SocketException::class)
	override fun setReuseAddress(on: Boolean) {
		parent.setReuseAddress(on)
	}

	@Synchronized
	@Throws(SocketException::class)
	override fun setSendBufferSize(size: Int) {
		parent.setSendBufferSize(size)
	}

	@Throws(SocketException::class)
	override fun setSoLinger(
		on: Boolean,
		linger: Int,
	) {
		parent.setSoLinger(on, linger)
	}

	@Synchronized
	@Throws(SocketException::class)
	override fun setSoTimeout(timeout: Int) {
		parent.setSoTimeout(timeout)
	}

	@Throws(SocketException::class)
	override fun setTcpNoDelay(on: Boolean) {
		parent.setTcpNoDelay(on)
	}

	@Throws(SocketException::class)
	override fun setTrafficClass(tc: Int) {
		parent.setTrafficClass(tc)
	}

	@Throws(IOException::class)
	override fun shutdownInput() {
		parent.shutdownInput()
	}

	@Throws(IOException::class)
	override fun shutdownOutput() {
		parent.shutdownOutput()
	}

	override fun toString(): String = parent.toString()
}
